src/runtime-rs/crates/hypervisor/src/qemu/cmdline_generator.rs

```diff
@@ -1178,6 +1183,133 @@ impl ToQemuParams for DeviceIntelIommu {
     }
 }
 
+// Qemu provides methods and types for managing QEMU instances.
+// To manage a qemu instance after it has been launched you need
+// to pass the -qmp option during launch requesting the qemu instance
+// to create a QMP unix domain manageent socket, e.g.,
+// -qmp unix:/tmp/qmp-socket,server=on,wait=off.
+#[derive(Debug, Default, PartialEq)]
+pub enum MonitorProtocol {
+    // Socket using a human-friendly text-based protocol.
+    Hmp,
+
+    // Socket using a richer json-based protocol.
+    #[default]
+    Qmp,
+
+    // Same as Qmp with pretty json formatting.
+    QmpPretty,
+}
+
+impl MonitorProtocol {
+    pub fn new(proto: &str) -> Self {
+        match proto {
+            "hmp" => MonitorProtocol::Hmp,
+            "qmp-pretty" => MonitorProtocol::QmpPretty,
+            _ => MonitorProtocol::Qmp,
+        }
+    }
+}
+
+impl ToString for MonitorProtocol {
+    fn to_string(&self) -> String {
+        match *self {
+            MonitorProtocol::Hmp => "hmp".to_string(),
+            MonitorProtocol::QmpPretty => "qmp-pretty".to_string(),
+            _ => "qmp".to_string(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct QmpSocket {
+    // type is the socket type (e.g. "unix").
+    pub socket_type: String,
+    // protocol to be used on the socket.
+    pub protocol: MonitorProtocol,
+    // QMP listener file descriptor to be passed to qemu
+    pub fd: Option<File>,
+    // name is the socket name.
+    pub name: String,
+    // server tells if this is a server socket.
+    pub server: bool,
+    // nowait tells if qemu should block waiting for a client to connect.
+    pub nowait: bool,
+}
+
+impl QmpSocket {
+    fn new(sid: &str) -> Result<Self> {
+        let sock_path = [KATA_PATH, sid, QMP_SOCKET_FILE].join("/");
+        let listener = UnixListener::bind(sock_path).context("unix listener bind failed.")?;
+
+        let raw_fd = listener.into_raw_fd();
+        clear_cloexec(raw_fd).context("clearing unix listenser O_CLOEXEC failed")?;
+        let sock_file = unsafe { File::from_raw_fd(raw_fd) };
+
+        Ok(QmpSocket {
+            socket_type: QMP_SOCKET_TYPE.to_owned(),
+            protocol: MonitorProtocol::new("qmp"),
+            fd: Some(sock_file),
+            name: QMP_SOCKET_FILE.to_owned(),
+            server: true,
+            nowait: true,
+        })
+    }
+
+    // validate returns true if the QmpSocket structure is valid and complete.
+    fn validate(&self) -> bool {
+        if self.socket_type.as_str() != QMP_SOCKET_TYPE {
+            return false;
+        }
+
+        if self.name.is_empty() && self.fd.is_none() {
```

Comment:
I was wondering if `||` would be more to catch invalid cases.

Reply:
My bad. Sorry for lack of knowledge. `name` and `fd` should not exist together. Please ignore my comment.